subsys/fs/nvs/nvs.c

```diff
@@ -162,214 +553,189 @@ void _nvs_entry_sector_advance(struct nvs_fs *fs)
-		rc = nvs_flash_read(fs, hdr_addr, &head, sizeof(head));
+	k_mutex_lock(&fs->nvs_lock, K_FOREVER);
+
+	for (i = 0; i < fs->sector_count; i++) {
+		addr = i * fs->sector_size;
+		rc = _nvs_flash_read_data(fs, addr, &sector_hdr,
+					  sizeof(struct _nvs_sector_hdr));
 		if (rc) {
-			return rc;
+			SYS_LOG_DBG("Flash read error %d", rc);
+			goto err;
 		}
-		if (head.id == NVS_ID_EMPTY) {
-			return -ENOENT;
+		if (sector_hdr.fd_magic != fs->magic) {
+			continue;
+		}
+		sectors_in_use++;
+		if (sectors_in_use == 1) {
+			fs->sector_id = sector_hdr.fd_id;
+			last_sector = i;
+			continue;
 		}
-		if (head.id == entry->id) {
-			entry->len = head.len;
-			return 0;
+		if (NVS_ID_GT(sector_hdr.fd_id, fs->sector_id)) {
+			fs->sector_id = sector_hdr.fd_id;
+			last_sector = i;
 		}
-		adv_len = _nvs_entry_len_in_flash(fs, head.len);
-		_nvs_addr_advance(fs, &entry->data_addr, adv_len);
 	}
-}
-
 
-int nvs_get_last_entry(struct nvs_fs *fs, struct nvs_entry *entry)
-{
-	int rc;
-	struct nvs_entry latest;
-	struct _nvs_data_hdr head;
-	off_t hdr_addr;
-	u16_t adv_len;
-
-	rc = nvs_get_first_entry(fs, entry);
-	if (rc) {
-		return rc;
-	}
-	latest.id = entry->id;
-	latest.data_addr = entry->data_addr;
-	latest.len = entry->len;
-	while (1) {
-		hdr_addr = _nvs_head_addr_in_flash(fs, entry);
-		rc = nvs_flash_read(fs, hdr_addr, &head, sizeof(head));
+	if ((!sectors_in_use) || (fs->locked)) {
+		/* No valid sectors found or reinit locked fs */
+		SYS_LOG_DBG("(Re)Initializing sectors");
+		for (addr = 0; addr < (fs->sector_count * fs->sector_size);
+			addr += fs->sector_size) {
+			/* first check if used, only erase if it is */
+			if (_nvs_sector_is_used(fs, addr)) {
+				rc = _nvs_flash_erase(fs, addr,
+						      fs->sector_size);
+				if (rc) {
+					goto err;
+				}
+			}
+		}
+		rc = _nvs_sector_init(fs, 0);
 		if (rc) {
-			return rc;
+			goto err;
 		}
-		if (head.id == NVS_ID_EMPTY) {
-			entry->id = latest.id;
-			entry->data_addr = latest.data_addr;
-			entry->len = latest.len;
-			return 0;
+		fs->entry_sector = 0;
+		fs->locked = false;
+	} else {
+		/* Valid sectors in flash, read to set fs->write_location */
+
+		/* First we need to calculate the entry sector, calculations
+		 * on u8_t so carefull
+		 */
+		sector_jmp = sectors_in_use - 1;
+		if (sector_jmp <= last_sector) {
+			fs->entry_sector = last_sector - sector_jmp;
+		} else {
+			fs->entry_sector = fs->sector_count - sector_jmp
+					   + last_sector;
 		}
-		if (head.id == latest.id) {
-			latest.len = head.len;
-			latest.data_addr = entry->data_addr;
+
+		addr = fs->entry_sector * fs->sector_size;
+
+		while (1) {
+			rc = _nvs_flash_rd_len_id_start(fs, addr, &id,
+							&len_start);
+			if (rc) {
+				goto err;
+			}
+			if (id == NVS_ID_EMPTY) {
+				fs->write_location = addr;
+				break;
+			}
+			_nvs_addr_advance(fs, &addr, _nvs_flash_size(fs,
+					  len_start));
+			rc = _nvs_flash_rd_len_id_end(fs, addr, &id, &len_end);
+			if (len_start != len_end) {
```

Comment:
I think it would be worth to serve the case when this is about the last entry which was not entirely written to the flash (caused for instance by power down).

Reply:
Hi @nvlsianpu, I am changing this routine a bit, fs->write_location is now updated everytime the length comparison is ok. So in the case of an entry that was not completely written the initialization routine will give a fs that is read-only but does not include the latest (invalid) entry. It is then possible to retrieve all information that is valid.

Reply:
What you are changing sound fine to me. Otherwise what I was asking might be added in a future PR as well. I think this is very likely to have the last entry corrupted for some devices. I would be great If nvs would can fix this in place by performing customized sector rotation.